ochazuke/webhooks/helpers.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+"""Helpers methods for webhooks."""
+
+import hashlib
+import hmac
+import logging
+
+from datetime import datetime
+
+from ochazuke import app
+from ochazuke.models import db, Milestone, Label, Issue, Event
+
+
+def get_payload_signature(key, payload):
+    """Compute the payload signature given a key."""
+    mac = hmac.new(key, msg=payload, digestmod=hashlib.sha1)
+    return mac.hexdigest()
+
+
+def signature_check(key, post_signature, payload):
+    """Check the HTTP POST legitimacy."""
+    if post_signature.startswith('sha1='):
+        sha_name, signature = post_signature.split('=')
+    else:
+        return False
+    if not signature:
+        return False
+    # HMAC requires its key to be bytes, but data is strings.
+    hexmac = get_payload_signature(key, payload)
+    return compare_digest(hexmac, signature.encode('utf-8'))
+
+
+def compare_digest(x, y):
+    """Create a hmac comparison.
+
+    Approximates hmac.compare_digest (Py 2.7.7+) until we upgrade.
+    TODO: SEE IF THIS NEEDS TO BE ADAPTED FOR PY3.6
+    """
+    if not (isinstance(x, bytes) and isinstance(y, bytes)):
+        raise TypeError("both inputs should be instances of bytes")
+    if len(x) != len(y):
+        return False
+    result = 0
+    for a, b in zip(bytearray(x), bytearray(y)):
+        result |= a ^ b
+    return result == 0
+
+
+def is_github_hook(request):
+    """Validate the github webhook HTTP POST request."""
+    if request.headers.get('X-GitHub-Event') is None:
+        return False
+    post_signature = request.headers.get('X-Hub-Signature')
+    if post_signature:
+        # TODO: DETERMINE WHERE TO HIDE SECRET KEY
+        key = app.config['HOOK_SECRET_KEY']
+        return signature_check(key, post_signature, request.data)
+    return False
+
+
+def process_issue_event_info(payload):
+    """Extract information we need when handling webhook for issue events."""
+    # Extract the event-specific info and make our own timestamp
+    event_timestmp = datetime.utcnow().replace(microsecond=0).isoformat() + 'Z'
+    action = payload.get('action')
+    milestone = payload.get('issue')['milestone']
+    # If there is no milestone data, let title be None
+    if milestone is not None:
+        milestone_title = milestone.get('title')
+    else:
+        milestone_title = None
+    # Let details be None for opening/closing, but preserve old title on edits
+    if action in ['opened', 'closed', 'reopened', 'edited']:
+        changes = payload.get('changes')
+        if changes is not None:
```

Comment:
You could also just write `if change:`, it will fail on falsey stuff (like `None`).

Reply:
@laghee 

just to be sure. Do you know if there is a "changes" with content when `opened`, `reopened` or `closed` action are happening?

and when `edited` is there any chance that changes is not there?

Reply:
@karlcow Nope -- for `opened`, `reopened`, and `closed`, there is no `changes` key. If it's not present on `edited`, however, it would be because of some error (or future configuration change) on GitHub's end (https://developer.github.com/v3/activity/events/types/#issuesevent).